src/wbase/ChannelShared.cc

```diff
@@ -0,0 +1,337 @@
+    QSERV_LOGCONTEXT_QUERY_JOB(qId, jId);
+    assert(tData != nullptr);
+
+    // This lock may be held for a very long time.
+    lock_guard<mutex> const queueMtxLock(_queueMtx);
+    _transmitQueue.push(tData);
+
+    // If _lastRecvd is true, the last message has already been transmitted and
+    // this SendChannel is effectively dead.
+    bool reallyLast = _lastRecvd;
+    string idStr(makeIdStr(qId, jId));
+    if (_icPtr == nullptr) {
+        _icPtr = make_shared<util::InstanceCount>(to_string(qId) + "_SCS_LDB");
+    }
+
+    // If something bad already happened, just give up.
+    if (reallyLast || isDead()) {
+        // If there's been some kind of error, make sure that nothing hangs waiting
+        // for this.
+        LOGS(_log, LOG_LVL_WARN, "addTransmit getting messages after isDead or reallyLast " << idStr);
+        _lastRecvd = true;
+        return false;
+    }
+
+    // If lastIn is true, all tasks for this job have run to completion and
+    // finished building their transmit messages.
+    if (lastIn) {
+        reallyLast = true;
+    }
+    if (reallyLast || erred || cancelled) {
+        _lastRecvd = true;
+        LOGS(_log, LOG_LVL_DEBUG,
+             "addTransmit lastRecvd=" << _lastRecvd << " really=" << reallyLast << " erred=" << erred
+                                      << " cancelled=" << cancelled);
+    }
+
+    return _transmit(tMtxLock, queueMtxLock, erred, task);
+}
+
+bool ChannelShared::_transmit(lock_guard<mutex> const& tMtxLock, lock_guard<mutex> const& queueMtxLock,
+                              bool erred, Task::Ptr const& task) {
+    string idStr = "QID?";
+
+    // Result data is transmitted in messages containing data and headers.
+    // data - is the result data
+    // header - contains information about the next chunk of result data,
+    //          most importantly the size of the next data message.
+    //          The header has a fixed size (about 255 bytes)
+    // header_END - indicates there will be no more msg.
+    // msg - contains data and header.
+    // metadata - special xrootd buffer that can only be set once per ChannelShared
+    //            instance. It is used to send the first header.
+    // A complete set of results to the czar looks like
+    //    metadata[header_A] -> msg_A[data_A, header_END]
+    // or
+    //    metadata[header_A] -> msg_A[data_A, header_B]
+    //          -> msg_B[data_B, header_C] -> ... -> msg_X[data_x, header_END]
+    //
+    // Since you can't send msg_A until you know the size of data_B, you can't
+    // transmit until there are at least 2 msg in the queue, or you know
+    // that msg_A is the last msg in the queue.
+    // Note that the order of result rows does not matter, but data_B must come after header_B.
+    // Keep looping until nothing more can be transmitted.
+    while (_transmitQueue.size() >= 2 || _lastRecvd) {
```

Comment:
This code is 100% compatible with the current version (at the `main` branch) of `wbase::SendChannelShared`, which has:
```
bool SendChannelShared::_transmit(bool erred, Task::Ptr const& task) {
    string idStr = "QID?";

    // Result data is transmitted in messages containing data and headers.
    // data - is the result data
    // header - contains information about the next chunk of result data,
    //          most importantly the size of the next data message.
    //          The header has a fixed size (about 255 bytes)
    // header_END - indicates there will be no more msg.
    // msg - contains data and header.
    // metadata - special xrootd buffer that can only be set once per SendChannelShared
    //            instance. It is used to send the first header.
    // A complete set of results to the czar looks like
    //    metadata[header_A] -> msg_A[data_A, header_END]
    // or
    //    metadata[header_A] -> msg_A[data_A, header_B]
    //          -> msg_B[data_B, header_C] -> ... -> msg_X[data_x, header_END]
    //
    // Since you can't send msg_A until you know the size of data_B, you can't
    // transmit until there are at least 2 msg in the queue, or you know
    // that msg_A is the last msg in the queue.
    // Note that the order of result rows does not matter, but data_B must come after header_B.
    // Keep looping until nothing more can be transmitted.
    while (_transmitQueue.size() >= 2 || _lastRecvd) {
```
If you think the algorithm has to be reinforced then we could do it as a separate effort.
I just don't want to introduce any chances of running into a new failure mode that might result in worker lock-ups or query cancellation troubles.